README.md
# hex [![Docker Hub build][docker-hub-badge]][docker-hub]

> A hydrus API plugin to download ExH archives

![Screenshot of the userscript in action][screenshot]

hex is a tool that connects to the [hydrus][hydrus] client API and allows you
to download and import ExH galleries via the archive download method (instead
of scraping the images), alongside their tags and URL association.

A simple HTTP API allows you to send gallery URLs to hex that it will then
proceed to download and import automatically. A [userscript](hex.user.js) that
adds a hex toolbar to ExH gallery pages is included for convenience.

## Table of contents

+ [Install](#install)
  + [Installing with Docker](#installing-with-docker)
  + [Installing without Docker](#installing-without-docker)
  + [Dependencies](#dependencies)
  + [Updating](#updating)
    + [Updating with Docker](#updating-with-docker)
    + [Updating without Docker](#updating-without-docker)
+ [Usage](#usage)
  + [Running with Docker](#running-with-docker)
  + [Running without Docker](#running-without-docker)
  + [Configuration](#configuration)
  + [Userscript](#userscript)
  + [API](#api)
    + [General](#general)
    + [Routes](#routes)
      + [Base](#base)
      + [Settings](#settings)
      + [Import](#import)
+ [Maintainer](#maintainer)
+ [Contribute](#contribute)
+ [License](#license)

## Install

The recommended way to run is via [Docker][docker]. Basic instructions on how
to run without it are also provided.

### Installing with Docker

To use hex with Docker, you can simply pull the prebuilt image from
[Docker Hub][docker-hub]:

```zsh
user@local:~$ docker pull mtbl/hex
```

Alternatively, you can also build the image yourself. The user that is used
inside the container has UID `1000` and GID `1000` by default. You can adjust
this (e.g., to match your host UID/GID) by providing the arguments `USER_ID`
and `GROUP_ID` when making a build.

### Installing without Docker

To install without Docker, you can simply clone the repository and install
dependencies.

```zsh
user@local:~$ git clone https://github.com/imtbl/hex.git
user@local:~$ cd hex
user@local:hex$ yarn
```

### Dependencies

+ [hydrus][hydrus]
+ [Docker][docker] (when using Docker)
+ [Node.js][node-js] (when not using Docker)
+ [Yarn][yarn] (when not using Docker)
+ A [Puppeteer][puppeteer]-compatible browser that hex uses for navigating and
  parsing ExH; [browserless][browserless] is recommended as a headless solution

hex should work with both the latest LTS and the latest stable version of
Node.js. If you encounter any issues with either of those versions when not
using Docker, please [let me know][issues].

### Updating

This application follows [semantic versioning][semantic-versioning] and any
breaking changes that require additional attention will be released under a new
major version (e.g., `2.0.0`). Minor version updates (e.g., `1.1.0` or `1.2.0`)
are therefore always safe to simply install.

When necessary, this section will be expanded with upgrade guides for new major
versions.

#### Updating with Docker

Simply pull the latest Docker image to update:

```zsh
user@local:~$ docker pull mtbl/hex
```

#### Updating without Docker

If you chose not to use Docker, you can update via Git:

```zsh
user@local:hex$ git pull
user@local:hex$ yarn
```

## Usage

hex has a few specific requirements that you need to keep in mind in order to
use it, regardless of if you decide to run with or without Docker:

+ It requires a [Puppeteer][puppeteer]-compatible browser that is used for
  navigating and parsing ExH; [browserless][browserless] is recommended as a
  headless solution and used in the included [Docker Compose][docker-compose]
  example setup.
+ hex needs access to the hydrus client API. The access key needs to have the
  following permissions: `import files`, `add tags to files` and
  `add urls for processing`.
+ The import path needs to be accessible by both hex and hydrus, so you need to
  find a way to achieve this should you decide to run hex on a different
  machine. This can, for example, be achieved by using [sshfs][sshfs].

### Running with Docker

To make using Docker as easy as possible, a working
[Docker Compose][docker-compose] example setup is provided. To get started with
this example setup, simply duplicate `docker-compose.yml.example` as
`docker-compose.yml` and adjust the variables in the `environment` section as
described [here](#configuration).

Pay special attention to the variable `HEX_DOCKER_HOST_IMPORT_PATH`. This is
only required when using Docker and needs to contain the __absolute__ path to
the import directory on the host machine that is mounted as volume. This is the
path hydrus will access to import the files from when hex asks it
to.

Finally, start the containers:

```zsh
user@local:hex$ docker-compose up -d
```

### Running without Docker

To run without Docker, you will first need to duplicate the `.env.example` as
`.env` and adjust the variables as described [here](#configuration).

After that, you can start hex:

```zsh
user@local:hex$ yarn start
```

### Configuration

Configuration is done entirely via environment variables. Please pay special
attention to the instructions to prevent issues.

+ `HEX_PORT=8000`: the port hex is listening on.
+ `HEX_ACCESS_KEY=`: an arbitrary string used as access key for hex's API. Can
  be of any (reasonable) length, the only requirement being that it is set at
  all.
+ `HEX_BROWSER_WS_ENDPOINT=ws://localhost:3000`: the WebSocket endpoint of the
  [Puppeteer]-compatible browser hex needs to be able to connect to. __No__
  __trailing slashes.__
+ `HEX_BROWSER_USER_AGENT=Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/88.0.4324.96 Safari/537.36`:
  the user agent to use with Puppeteer.
+ `HEX_HYDRUS_BASE_URL=http://localhost:45869`: the hydrus client API base URL.
  __No trailing slashes.__
+ `HEX_HYDRUS_ALLOW_NEWER_API_VERSION=false`: setting this to `true` allows hex
  to connect to a newer hydrus client API versions than it officially supports.
  Enable this on your own risk to (possibly) be able to continue using hex if
  the hydrus API gets updated and a new hex release that reflects this is not
  out yet. Be aware that this can lead to hex not working at all or imports
  getting (partially) broken.
+ `HEX_HYDRUS_ACCESS_KEY=`: the hydrus client API access key hex uses to
  connect. Needs to have the following permissions: `import files`,
  `add tags to files` and `add urls for processing`.
+ `HEX_HYDRUS_TAG_SERVICE=my tags`: the hydrus tag service hex adds tags to.
+ `HEX_SKIP_IMPORT=false`: setting this to `true` downloads the archive and
  extracts it, but skips the hydrus import altogether.
+ `HEX_IMPORT_PATH=./import`: the path hex saves archive downloads at and
  extracts them in. Can be relative or absolute.
+ `HEX_DOCKER_HOST_IMPORT_PATH=`: only required when using Docker and needs to
  contain the __absolute__ path to the import directory on the host machine
  that is mounted as volume. The path has to be in a format your host operating
  system supports.
+ `HEX_SKIP_KNOWN_FILES=false`: setting this to `true` skips files hydrus knows
  about altogether (this will neither try to import them nor attempt to update
  their tags).
+ `HEX_DELETE_ARCHIVES_AFTER_IMPORT=true`: setting this to `false` will cause
  the extracted archives stored under `HEX_IMPORT_PATH` not to be deleted once
  the hydrus import finishes.
+ `HEX_SKIP_TAGS=false`: setting this to `true` will prevent any new tags from
  being added to imported files and disregards any other tag-related settings.
+ `HEX_BLACKLISTED_NAMESPACES=`: namespaces that are added here separated with
  `###` will be excluded from getting added to hydrus. E.g.,
  `artist###language###misc`. This only applies to tags sourced from ExH and
  the special `page:<page number>` tag that is added by default, not to tags
  added via `HEX_ADDITIONAL_TAGS`. In addition, if `HEX_NAMESPACE_REPLACEMENTS`
  is used and the replacement (but not the original) is a blacklisted
  namespace, it will still be added as well.
+ `HEX_NAMESPACE_REPLACEMENTS='artist|||creator###parody|||series###female|||###male|||###group|||###misc|||'`:
  namespaces that are added here in the format `<original>|||<replacement>` and
  separated with `###` will be replaced accordingly. Leaving out the
  replacement altogether (e.g., `###misc|||`) _unnamespaces_ them.
+ `HEX_ADDITIONAL_TAGS=`: additional tags to be added. Have to be provided in
  the format `<namespace>:<tag>` or simply `<tag>` (for unnamespaced tags) and
  separated with `###`.
+ `HEX_ADD_UNIQUE_IDENTIFIER_TAG=false`: setting this to true causes hex to
  add a tag in the form
  `<HEX_UNIQUE_IDENTIFIER_NAMESPACE>:<ExH gallery ID>-<page>`. This tag is
  intended to uniquely identify the position of an image inside an archive in
  case it is used across multiple ones (in which case it might have
  multiple/different `page` tags, making it hard to determine which one belongs
  to which archive).
+ `HEX_UNIQUE_IDENTIFIER_NAMESPACE=unique`: the namespace to use for the unique
  identifier tag when `HEX_ADD_UNIQUE_IDENTIFIER_TAG` is `true`.

### Userscript

To make using hex as comfortable as possible, a [userscript](hex.user.js) that
adds a hex toolbar to ExH gallery pages is included. When you first open ExH
with the userscript enabled, it will prompt you for the hex base URL and the
access key. You can adjust these at any point in the settings, but be sure to
refresh any open ExH page after doing so (as the changes will not be reflected
on a page that had already been loaded before changing base URL or access key).
__Please keep in mind that the access key is stored in plaintext and that__
__anyone with access to the browser can read it.__

### API

#### General

Request and response bodies are always in JSON format. The `Authorization`
header in the format `Authorization: Bearer <HEX_ACCESS_KEY>` is used to
authenticate for all routes except the base route (`/`).

Requests with missing or malformed parameters will be responded with an error
in the following format and error code `400`:

```json5
{
  "error": <field name>
}
```

#### Routes

##### Base

Responds with the version number and the API version number of the hex
installation. The API version number will increase by 1 every time an existing
API endpoint is modified in a way it behaves differently than before or
removed altogether.

__Route:__ `GET /`

__Response on success:__

```json5
{
  "hex": {
    "version": <version number of hex installation>,
    "apiVersion>": <API version number of hex installation>
  }
}
```

##### Settings

Responds with a number of default settings configured via environment
variables. These settings can then be overridden per import.

__Route:__ `GET /settings`

__Response on success:__

```json5
{
  "settings": {
    "skipImport": <boolean indicating if hydrus imports should be skipped>,
    "skipKnownFiles": <boolean indicating if known files should be skipped>,
    "deleteArchivesAfterImport": <boolean indicating if archives should be deleted after import>,
    "skipTags": <boolean indicating if adding tags should be skipped>,
    "blacklistedNamespaces": <array of blacklisted namespaces>,
    "namespaceReplacements": <object where the key is the original and the value the replacement namespace>,
    "additionalTags": <array of additional tags>
  }
}
```

##### Import

Used to send ExH gallery URLs to hex for processing.

__Route:__ `POST /import`

__Request body:__

```json5
{
  "cookies": <ExH `Cookie` header>,
  "url": <ExH gallery URL to be processed>,
  "skipImport": <boolean indicating if hydrus imports should be skipped>, // optional, if not provided, the default will be used
  "skipKnownFiles": <boolean indicating if known files should be skipped>, // optional, if not provided, the default will be used
  "deleteArchivesAfterImport": <boolean indicating if archives should be deleted after import>, // optional, if not provided, the default will be used
  "skipTags": <boolean indicating if adding tags should be skipped>, // optional, if not provided, the default will be used
  "blacklistedNamespaces": <blacklisted namespaces in the same format as `HEX_BLACKLISTED_NAMESPACES`>, // optional, if not provided, the default will be used
  "namespaceReplacements": <namespace replacements in the same format as `HEX_NAMESPACE_REPLACEMENTS`>, // optional, if not provided, the default will be used
  "additionalTags": <additional tags in the same format as `HEX_ADDITIONAL_TAGS`> // optional, if not provided, the default will be used
}
```

__Response on success:__

```json5
{
  "import": <ExH gallery URL> // does not indicate success, only that the processing of the gallery has been started
}
```

## Maintainer

[imtbl][maintainer]

## Contribute

You are welcome to help out!

[Open an issue][issues] or submit a pull request.

## License

[AGPLv3](LICENSE) © imtbl

[hydrus]: http://hydrusnetwork.github.io/hydrus
[docker]: https://www.docker.com/
[docker-hub]: https://hub.docker.com/r/mtbl/hex/
[node-js]: https://nodejs.org/en/
[yarn]: https://yarnpkg.com/
[puppeteer]: https://pptr.dev/
[browserless]: https://github.com/browserless/chrome
[semantic-versioning]: https://semver.org/
[docker-compose]: https://docs.docker.com/compose/
[sshfs]: https://github.com/libfuse/sshfs

[screenshot]: https://github.com/imtbl/hex/raw/master/media/screenshot.png

[docker-hub-badge]: https://img.shields.io/docker/cloud/automated/mtbl/hex.svg

[maintainer]: https://github.com/imtbl
[issues]: https://github.com/imtbl/hex/issues/new


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to
[Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.14.0] - 2021-02-17

### Changed

+ Made it so that user agent and cookies are also used for the actual archive
  download
+ Made the namespace that is used for the unique identifier tag configurable

### Fixed

+ Fixed the hex toolbar (that is added via userscript) not appearing on gallery
  pages and instead incorrectly displaying on all other ExH pages

## [1.13.0] - 2021-02-02

### Added

+ Added the ability to use a custom user agent with Puppeteer

## [1.12.0] - 2020-12-19

### Changed

+ Updated the hydrus API version to `15`

## [1.11.0] - 2020-10-03

### Changed

+ Updated the hydrus API version to `14`

## [1.10.0] - 2020-06-29

### Changed

+ Updated the hydrus API version to `13`

## [1.9.0] - 2020-06-17

### Changed

+ Updated the hydrus API version to `12`

## [1.8.0] - 2020-05-22

### Changed

+ Maintenance release, no feature changes

## [1.7.0] - 2020-04-22

### Added

+ Added more log output about the currently worked on file

### Changed

+ Updated dependencies

## [1.6.0] - 2020-03-07

### Changed

+ Updated dependencies

### Fixed

+ Corrected the inability to extract large archives by using a stream instead
  of a buffer

## [1.5.1] - 2020-01-23

### Fixed

+ Errors occuring during navigation to the download page are now handled

## [1.5.0] - 2020-01-22

### Added

+ Added the ability to add a special identifier tag to uniquely determine the
  position of an image inside an archive

### Changed

+ Updated dependencies

## [1.4.0] - 2020-01-12

### Removed

+ Removed the ability to define custom UID and GID for the Docker container at
  container creation due to several issues arising from that

## [1.3.1] - 2020-01-07

### Fixed

+ Fixed access control headers not being added after restana upgrade

## [1.3.0] - 2020-01-07

### Added

+ Added the ability to define custom UID and GID for the Docker container at
  container creation

### Changed

+ Updated dependencies

## [1.2.0] - 2019-12-26

### Changed

+ Adjusted the workaround implemented for the discrepancy between the
  `/add_urls/associate_url` and `/add_urls/get_url_files` routes so it makes
  sure the normalized URL does not end with a `/`

### Fixed

+ Added missing example namespace replacements to Docker Compose example
  configuration

## [1.1.0] - 2019-12-24

### Added

+ Added additional example namespace replacements
+ Added the ability to skip adding tags

### Changed

+ Reworked API and made settings overridable
+ Updated and expanded userscript
+ Updated dependencies

### Fixed

+ Fixed error when not providing `HEX_DOCKER_HOST_IMPORT_PATH`
+ Added a workaround for the discrepancy between the `/add_urls/associate_url`
  and `/add_urls/get_url_files` hydrus API routes in regards to normalization

## [1.0.1] - 2019-12-19

### Fixed

+ Removed unnecessary Travis CI configuration

## 1.0.0 - 2019-12-19

### Added

+ Initial release

[Unreleased]: https://github.com/imtbl/hex/compare/1.14.0...develop
[1.14.0]: https://github.com/imtbl/hex/compare/1.13.0...1.14.0
[1.13.0]: https://github.com/imtbl/hex/compare/1.12.0...1.13.0
[1.12.0]: https://github.com/imtbl/hex/compare/1.11.0...1.12.0
[1.11.0]: https://github.com/imtbl/hex/compare/1.10.0...1.11.0
[1.10.0]: https://github.com/imtbl/hex/compare/1.9.0...1.10.0
[1.9.0]: https://github.com/imtbl/hex/compare/1.8.0...1.9.0
[1.8.0]: https://github.com/imtbl/hex/compare/1.7.0...1.8.0
[1.7.0]: https://github.com/imtbl/hex/compare/1.6.0...1.7.0
[1.6.0]: https://github.com/imtbl/hex/compare/1.5.1...1.6.0
[1.5.1]: https://github.com/imtbl/hex/compare/1.5.0...1.5.1
[1.5.0]: https://github.com/imtbl/hex/compare/1.4.0...1.5.0
[1.4.0]: https://github.com/imtbl/hex/compare/1.3.1...1.4.0
[1.3.1]: https://github.com/imtbl/hex/compare/1.3.0...1.3.1
[1.3.0]: https://github.com/imtbl/hex/compare/1.2.0...1.3.0
[1.2.0]: https://github.com/imtbl/hex/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/imtbl/hex/compare/1.0.1...1.1.0
[1.0.1]: https://github.com/imtbl/hex/compare/1.0.0...1.0.1


hex.user.js
// ==UserScript==
// @name hex
// @version 1.14.0
// @author imtbl
// @description Userscript for hex, a hydrus API plugin to download ExH archives
// @website https://github.com/imtbl/hex
// @updateURL https://github.com/imtbl/hex/raw/master/hex.user.js
// @downloadURL https://github.com/imtbl/hex/raw/master/hex.user.js
// @supportURL https://github.com/imtbl/hex/issues/new
// @match *://exhentai.org/*
// @grant GM_setValue
// @grant GM_getValue
// @grant GM_registerMenuCommand
// ==/UserScript==

let hexBaseUrl = window.GM_getValue('hexBaseUrl', '')
let hexAccessKey = window.GM_getValue('hexAccessKey', '')

const fetchOrPrompt = (value, prompt, key, defaultValue = '') => {
  if (!value) {
    value = window.prompt(`${prompt} not set, please enter it:`, defaultValue)

    window.GM_setValue(key, value)
  }

  return value
}

const promptAndChangeStoredValue = (value, prompt, key) => {
  value = window.prompt(`Change ${prompt}:`, value)

  window.GM_setValue(key, value)
}

const changeBaseUrl = () => {
  promptAndChangeStoredValue(hexBaseUrl, 'hex base URL', 'hexBaseUrl')
}

const changeAccessKey = () => {
  promptAndChangeStoredValue(hexAccessKey, 'hex access key', 'hexAccessKey')
}

window.GM_registerMenuCommand('Change hex base URL', changeBaseUrl)
window.GM_registerMenuCommand('Change hex access key', changeAccessKey)

hexBaseUrl = fetchOrPrompt(
  hexBaseUrl, 'hex base URL', 'hexBaseUrl', 'http://localhost:8000'
)
hexAccessKey = fetchOrPrompt(hexAccessKey, 'hex access key', 'hexAccessKey')

const elementFromHtml = html => {
  const template = document.createElement('template')
  html = html.trim()
  template.innerHTML = html

  return template.content.firstChild
}

const getHexInfo = () => {
  return window.fetch(`${hexBaseUrl}`)
}

const getHexSettings = () => {
  return window.fetch(`${hexBaseUrl}/settings`, {
    headers: {
      Authorization: `Bearer ${hexAccessKey}`
    }
  })
}

const sendToHex = data => {
  return window.fetch(`${hexBaseUrl}/import`, {
    method: 'post',
    headers: {
      Authorization: `Bearer ${hexAccessKey}`,
      'Content-Type': 'application/json'
    },
    body: JSON.stringify(Object.fromEntries(data))
  })
}

const getDelimitedString = arr => {
  return arr.length
    ? arr.join('###')
    : ''
}

const getDelimitedNamespaceReplacements = namespaceReplacements => {
  const delimitedNamespaceReplacements = []

  Object.entries(namespaceReplacements).forEach(([key, value]) => {
    delimitedNamespaceReplacements.push(`${key}|||${value}`)
  })

  return delimitedNamespaceReplacements.join('###')
}

const start = async () => {
  if (!document.getElementById('gd5')) {
    return
  }

  let info, settings

  try {
    let infoResponse = await getHexInfo()
    infoResponse = await infoResponse.json()
    info = infoResponse.hex
  } catch (err) {
    console.error(err)

    return
  }

  if (info.apiVersion !== 2) {
    window.alert(
      'This userscript supports API version 2, while your hex ' +
      `installation is using version ${info.apiVersion}. Please adjust ` +
      'your installation or change the userscript manually.'
    )

    return
  }

  try {
    let settingsResponse = await getHexSettings()
    settingsResponse = await settingsResponse.json()
    settings = settingsResponse.settings
  } catch (err) {
    console.error(err)

    return
  }

  const hexBar = elementFromHtml(
    `<div class="hex-bar">
      <style>
        .hex-bar {
          align-items: center;
          background-color: #4f535b;
          border-bottom: 1px solid #000;
          box-sizing: border-box;
          display: flex;
          height: 50px;
          left: 0;
          padding: 0 20px;
          position: fixed;
          top: 0;
          width: 100%;
          z-index: 1000;
        }

        .hex-bar__title {
          font-size: 1rem;
          font-weight: 700;
          margin-right: auto;
        }

        .hex-bar__form,
        .hex-bar__form__settings {
          align-items: center;
          display: flex;
        }

        .hex-bar__form {
          margin-left: 20px;
        }

        .hex-bar__form__settings label {
          margin-left: 10px;
        }

        .hex-bar__form__settings label:first-child {
          margin-left: 0;
        }

        .hex-bar__form__settings input[type="checkbox"] {
          top: 0;
        }

        .hex-bar__form__settings input[type="text"] {
          width: 150px;
        }

        .hex-bar__form__download {
          margin-left: 20px;
        }

        .hex-bar__form__download button {
          -moz-appearance: none;
          -webkit-appearance: none;
          background: #5fa9cf;
          border: 1px solid #5fa9cf;
          border-radius: 5px;
          color: #f1f1f1;
          cursor: pointer;
          outline: 0;
          padding: 8px;
        }

        .hex-bar__form__download button:hover {
          background-color: #32789c;
          border-color: #32789c;
        }
      </style>
      <div class="hex-bar__title">hex</div>
      <form id="hex-form" class="hex-bar__form" action="#">
        <div class="hex-bar__form__settings">
          <label>Skip import <input type="checkbox" name="skipImport"${settings.skipImport ? ' checked' : ''}></label>
          <label>Skip known files <input type="checkbox" name="skipKnownFiles"${settings.skipKnownFiles ? ' checked' : ''}></label>
          <label>Delete archives after import <input type="checkbox" name="deleteArchivesAfterImport"${settings.deleteArchivesAfterImport ? ' checked' : ''}></label>
          <label>Add unique identifier tag <input type="checkbox" name="addUniqueIdentifierTag"${settings.addUniqueIdentifierTag ? ' checked' : ''}></label>
          <label>Skip tags <input type="checkbox" name="skipTags"${settings.skipTags ? ' checked' : ''}></label>
          <label>
            Blacklisted namespaces<br>
            <input type="text" name="blacklistedNamespaces" placeholder="Blacklisted namespaces" value="${getDelimitedString(settings.blacklistedNamespaces)}">
          </label>
          <label>
            Namespace replacements<br>
            <input type="text" name="namespaceReplacements" placeholder="Namespace replacements" value="${getDelimitedNamespaceReplacements(settings.namespaceReplacements)}">
          </label>
          <label>
            Additional tags<br>
            <input type="text" name="additionalTags" placeholder="Additional tags" value="${getDelimitedString(settings.additionalTags)}">
          </label>
          <label>
            Unique identifier namespace<br>
            <input type="text" name="uniqueIdentifierNamespace" placeholder="Unique identifier namespace" value="${settings.uniqueIdentifierNamespace}">
          </label>
        </div>
        <div class="hex-bar__form__download">
          <button id="hex-download-button" type="submit">Download</button>
        </div>
      </form>
    </div>`
  )

  document.getElementById('nb').style.cssText = 'margin-top: 60px;'

  document.body.insertAdjacentElement(
    'afterbegin', hexBar
  )

  const hexForm = document.getElementById('hex-form')
  const hexSubmitButton = document.getElementById('hex-download-button')

  hexForm.addEventListener('submit', event => {
    event.preventDefault()

    const data = new window.FormData(hexForm)

    data.set('skipImport', data.has('skipImport'))
    data.set('skipKnownFiles', data.has('skipKnownFiles'))
    data.set(
      'deleteArchivesAfterImport',
      data.has('deleteArchivesAfterImport')
    )
    data.set('addUniqueIdentifierTag', data.has('addUniqueIdentifierTag'))
    data.set('skipTags', data.has('skipTags'))

    data.set('cookies', document.cookie)
    data.set('url', window.location.href)

    sendToHex(data)
      .then(response => Promise.all([response.ok, response.json()]))
      .then(([responseOk, body]) => {
        if (!responseOk) {
          throw new Error(body.error)
        }

        hexSubmitButton.textContent = 'Started'
        hexSubmitButton.style.cssText =
          'background: #6a936d; border-color: #617c63;'
      })
      .catch(err => {
        console.error('Error trying to send this gallery to hex:\n', err)

        hexSubmitButton.textContent = 'Error'
        hexSubmitButton.style.cssText =
          'background: #9e2720; border-color: #9e2720;'
      })
  })
}

(() => {
  if (
    document.readyState === 'complete' ||
    document.readyState === 'loaded' ||
    document.readyState === 'interactive'
  ) {
    start()

    return
  }

  window.addEventListener('DOMContentLoaded', () => {
    start()
  })
})()
